Makefile

```diff
@@ -219,4 +220,16 @@ image/build:
 image/push:
 	${DOCKER} push ${IMG}:$(IMG_VERSION)
 
+# build and push multi-arch docker image
+PLATFORMS ?= linux/amd64,linux/s390x,linux/ppc64le
+.PHONY: image/buildx
+image/buildx: ## Build and push docker image for cross-platform support
+        # copy existing Dockerfile and insert --platform=${BUILDPLATFORM} into Dockerfile.cross, and preserve the original Dockerfile
+	sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' ./Dockerfiles/Dockerfile > ./Dockerfiles/Dockerfile.cross
```

Comment:
```suggestion
	sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' ./Dockerfile > ./Dockerfile.cross
```

Reply:
the above will at least kick-start the image/buildx, but later fails with:

```
make image/buildx
sed -e '1 s/\(^FROM\)/FROM --platform=\$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$\{BUILDPLATFORM\}/' ./Dockerfiles/Dockerfile > ./Dockerfiles/Dockerfile.cross
/bin/sh: ./Dockerfiles/Dockerfile.cross: No such file or directory
make: *** [image/buildx] Error 1
model-registry % sed -e '1 s/\(^FROM\)/FROM --platform=\$$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$$\{BUILDPLATFORM\}/' ./Dockerfiles/Dockerfile > ./Dockerfiles/Dockerfile.cross
zsh: no such file or directory: ./Dockerfiles/Dockerfile.cross
model-registry % code .
model-registry % make image/buildx                                                                                                                                                            
sed -e '1 s/\(^FROM\)/FROM --platform=\$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$\{BUILDPLATFORM\}/' ./Dockerfile > ./Dockerfile.cross
docker buildx create --name project-v3-builder
Failed to initialize: unable to resolve docker endpoint: context "colima": context not found: open /Users/mmortari/.docker/contexts/meta/f24fd3749c1368328e2b149bec149cb6795619f244c5b584e844961215dadd16/meta.json: no such file or directory
make: [image/buildx] Error 1 (ignored)
docker buildx use project-v3-builder
ERROR: failed to find instance "project-v3-builder": open /Users/mmortari/.docker/buildx/instances/project-v3-builder: no such file or directory
make: *** [image/buildx] Error 1
model-registry % export DOCKER_HOST=unix:///Users/mmortari/.local/share/containers/podman/machine/qemu/podman.sock
export TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE="/var/run/docker.sock"
model-registry % make image/buildx                                                                                
sed -e '1 s/\(^FROM\)/FROM --platform=\$\{BUILDPLATFORM\}/; t' -e ' 1,// s//FROM --platform=\$\{BUILDPLATFORM\}/' ./Dockerfile > ./Dockerfile.cross
docker buildx create --name project-v3-builder
project-v3-builder
docker buildx use project-v3-builder
docker buildx build --push --platform=linux/amd64,linux/s390x,linux/ppc64le --tag kubeflow/model-registry:main -f Dockerfile.cross .
[+] Building 173.8s (21/35)                                                                                                                                                                                                                docker-container:project-v3-builder
 => [internal] booting buildkit                                                                                                                                                                                                                                           2.0s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                                                                                                                        1.5s
 => => creating container buildx_buildkit_project-v3-builder0                                                                                                                                                                                                             0.5s
 => [internal] load build definition from Dockerfile.cross                                                                                                                                                                                                                0.0s
 => => transferring dockerfile: 1.92kB                                                                                                                                                                                                                                    0.0s
 => [linux/s390x internal] load metadata for registry.access.redhat.com/ubi8/ubi-minimal:8.8                                                                                                                                                                              1.9s
 => [linux/amd64 internal] load metadata for registry.access.redhat.com/ubi8/ubi-minimal:8.8                                                                                                                                                                              1.9s
 => [linux/ppc64le internal] load metadata for registry.access.redhat.com/ubi8/ubi-minimal:8.8                                                                                                                                                                            1.9s
 => [linux/arm64 internal] load metadata for registry.access.redhat.com/ubi8/go-toolset:1.19                                                                                                                                                                              2.9s
 => [internal] load .dockerignore                                                                                                                                                                                                                                         0.0s
 => => transferring context: 52B                                                                                                                                                                                                                                          0.0s
 => [linux/s390x stage-1 1/3] FROM registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                               31.6s
 => => resolve registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                                                    0.0s
 => => sha256:8737e1657f05079caa8c726fb829e674f98ab5a3dd612e1bbd5a3b46b095e360 37.55MB / 37.55MB                                                                                                                                                                         31.1s
 => => extracting sha256:8737e1657f05079caa8c726fb829e674f98ab5a3dd612e1bbd5a3b46b095e360                                                                                                                                                                                 0.5s
 => [internal] load build context                                                                                                                                                                                                                                         0.3s
 => => transferring context: 6.77MB                                                                                                                                                                                                                                       0.3s
 => [linux/arm64 builder  1/21] FROM registry.access.redhat.com/ubi8/go-toolset:1.19@sha256:dd16d34f7e4fae45a620ba8fdde378d0853314e4e0f4998658fd81628b36f0d6                                                                                                             92.4s
 => => resolve registry.access.redhat.com/ubi8/go-toolset:1.19@sha256:dd16d34f7e4fae45a620ba8fdde378d0853314e4e0f4998658fd81628b36f0d6                                                                                                                                    0.0s
 => => sha256:b953f5a6908d8aa51e2200fe3aa4b40102eb758be1d56e7dd3bec2a9cb85662e 162.55MB / 162.55MB                                                                                                                                                                       90.2s
 => => sha256:b8104c5f40fdcf0549bf9ad8b9ea4e5d3fabcc446f920cb976986f255e884a42 143.61MB / 143.61MB                                                                                                                                                                       70.0s
 => => sha256:8550ec08a06cf30baf540ef95dcc7334958355ef556b38eca4fd67cd68de5af1 18.17MB / 18.17MB                                                                                                                                                                         13.0s
 => => sha256:e1bb0572465a9e03d7af5024abb36d7227b5bf133c448b54656d908982127874 76.88MB / 76.88MB                                                                                                                                                                         43.1s
 => => extracting sha256:e1bb0572465a9e03d7af5024abb36d7227b5bf133c448b54656d908982127874                                                                                                                                                                                 1.2s
 => => extracting sha256:8550ec08a06cf30baf540ef95dcc7334958355ef556b38eca4fd67cd68de5af1                                                                                                                                                                                 0.3s
 => => extracting sha256:b8104c5f40fdcf0549bf9ad8b9ea4e5d3fabcc446f920cb976986f255e884a42                                                                                                                                                                                 2.2s
 => => extracting sha256:b953f5a6908d8aa51e2200fe3aa4b40102eb758be1d56e7dd3bec2a9cb85662e                                                                                                                                                                                 2.2s
 => [linux/ppc64le stage-1 1/3] FROM registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                             17.1s
 => => resolve registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                                                    0.0s
 => => sha256:655f24e7656ea1c1a2711179ea0c6d6f7dac72ba7d2d3fd34c36d062f00f74d8 44.03MB / 44.03MB                                                                                                                                                                         16.5s
 => => extracting sha256:655f24e7656ea1c1a2711179ea0c6d6f7dac72ba7d2d3fd34c36d062f00f74d8                                                                                                                                                                                 0.6s
 => [linux/amd64 stage-1 1/3] FROM registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                               31.4s
 => => resolve registry.access.redhat.com/ubi8/ubi-minimal:8.8@sha256:b93deceb59a58588d5b16429fc47f98920f84740a1f2ed6454e33275f0701b59                                                                                                                                    0.0s
 => => sha256:dc35b837139a95d1b9f7f7b0435a024a74ab972416bdc248f3f608c9f917a753 39.31MB / 39.31MB                                                                                                                                                                         30.8s
 => => extracting sha256:dc35b837139a95d1b9f7f7b0435a024a74ab972416bdc248f3f608c9f917a753                                                                                                                                                                                 0.5s
 => [linux/arm64 builder  2/21] WORKDIR /workspace                                                                                                                                                                                                                        0.3s
 => [linux/arm64 builder  3/21] COPY [go.mod, go.sum, ./]                                                                                                                                                                                                                 0.0s
 => [linux/arm64 builder  4/21] RUN go mod download                                                                                                                                                                                                                      17.6s
 => [linux/arm64 builder  5/21] RUN yum remove -y nodejs npm                                                                                                                                                                                                              0.8s
 => [linux/arm64 builder  6/21] RUN yum module -y reset nodejs                                                                                                                                                                                                            6.2s
 => [linux/arm64 builder  7/21] RUN yum module -y enable nodejs:18                                                                                                                                                                                                        0.4s
 => [linux/arm64 builder  8/21] RUN yum install -y nodejs npm java-11                                                                                                                                                                                                    50.6s
 => [linux/arm64 builder  9/21] COPY [Makefile, main.go, .openapi-generator-ignore, openapitools.json, ./]                                                                                                                                                                0.1s
 => ERROR [linux/arm64 builder 10/21] RUN set -ex    ; ARCH=$(uname -m) ; echo $ARCH     ; if [ "$ARCH" == "s390x" ] ; then ARCH="s390_64" ; elif [ "$ARCH" == "ppc64le" ] ; then ARCH="ppcle_64" ; fi     ; wget -q https://github.com/protocolbuffers/protobuf/release  0.4s
------
 > [linux/arm64 builder 10/21] RUN set -ex    ; ARCH=$(uname -m) ; echo $ARCH     ; if [ "$ARCH" == "s390x" ] ; then ARCH="s390_64" ; elif [ "$ARCH" == "ppc64le" ] ; then ARCH="ppcle_64" ; fi     ; wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-$ARCH.zip -O protoc.zip     ; unzip -q protoc.zip     ; bin/protoc --version     ; rm protoc.zip     ;:
0.037 ++ uname -m
0.039 + ARCH=aarch64
0.039 + echo aarch64
0.039 + '[' aarch64 == s390x ']'
0.039 + '[' aarch64 == ppc64le ']'
0.039 + wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-aarch64.zip -O protoc.zip
0.039 aarch64
------
Dockerfile.cross:24
--------------------
  23 |     # Download protoc compiler v24.3
  24 | >>> RUN set -ex\
  25 | >>>     ; ARCH=$(uname -m) ; echo $ARCH \
  26 | >>>     ; if [ "$ARCH" == "s390x" ] ; then ARCH="s390_64" ; elif [ "$ARCH" == "ppc64le" ] ; then ARCH="ppcle_64" ; fi \
  27 | >>>     ; wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-$ARCH.zip -O protoc.zip \
  28 | >>>     ; unzip -q protoc.zip \
  29 | >>>     ; bin/protoc --version \
  30 | >>>     ; rm protoc.zip \
  31 | >>>     ;
  32 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c set -ex    ; ARCH=$(uname -m) ; echo $ARCH     ; if [ \"$ARCH\" == \"s390x\" ] ; then ARCH=\"s390_64\" ; elif [ \"$ARCH\" == \"ppc64le\" ] ; then ARCH=\"ppcle_64\" ; fi     ; wget -q https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-$ARCH.zip -O protoc.zip     ; unzip -q protoc.zip     ; bin/protoc --version     ; rm protoc.zip     ;" did not complete successfully: exit code: 8
make: [image/buildx] Error 1 (ignored)
docker buildx rm project-v3-builder
project-v3-builder removed
rm Dockerfile.cross
```

Reply:
I have tested with s390x , pp64cle & amd64. its failing because above testing has been done for arm64. arm64 support is not there in Makefile.

Reply:
I'm not understanding how it can work without https://github.com/kubeflow/model-registry/pull/62/files#r1562588431

can you elaborate?

Reply:
There is small confusion from my end. I have put my changes on 1 week old code . I will rebase my change on to the latest code & raise again. As part of that your abnove comments will be addressed

Reply:
No problem, thank you!

Reply:
Handle it in the latest changes . @tarilabs please review